### ISSUE TITLE
Float16 support for log_uniform_candidate_sampler/uniform_candidate_sampler

### ISSUE THREAD

**Issue by Santosh-Gupta** [tensorflow/tensorflow#23086]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.11.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Float16 architectures are promising for significantly increase the model size / training speed using the same amount of memory. So I am requesting Float16 support for `log_uniform_candidate_sampler` and `uniform_candidate_sampler`. Currently, in the returned tuple,  `true_expected_count` and `sampled_expected_count` as returned as float32. This causes issues when using `sampled_softmax_loss` since everything needs to be in the same datatype. 

I was able to find a hack work around

https://stackoverflow.com/questions/52711895/how-to-run-define-tensorflow-graph-were-all-variables-are-in-float16-instead-ins

However, this actually increases the training speed; I am not sure why (more details below)

I have tried to develop a replacement function, however, I was unable to interpret the machine generated code where `log_uniform_candidate_sampler` and `uniform_candidate_sampler` are defined. I describe my attempt to interpret the code here

https://stackoverflow.com/questions/52734709/how-to-interpret-this-machine-generated-python-code

**Will this change the current api? How?**

Yes, most likely the best way is to have an optional arg for what datatype the user would like use, ie datatype=tf.float16. Otherwise defaults to float32

**Who will benefit with this feature?**

Those who are developing embedding items (ie Word2Vec) using float16 embeddings. 

**Any Other info.**

>Have I written custom code

Here is the code I used, for convenience it's in a Google Colab notebook so you can just run the code in the notebook. 

https://drive.google.com/open?id=1bnlSfezPqtwapvURRTAqGR4mkcCl9QZl

I wrote a custom function to replace `true_expected_count` and `sampled_expected_count` generated from `uniform_candidate_sampler` because I wanted to isolate the issue if the slow-down is due to using float16 instead of float32. The specific code to replace those variables is  

```
    LogUniformCandidateSampler = namedtuple("namedtuple", ["sampled_candidates", "true_expected_count", "sampled_expected_count"]) 
    sampled_values = tf.nn.uniform_candidate_sampler(
          true_classes=tf.cast(train_labels, tf.int64), num_sampled=num_sampled,
          num_true=1,
          unique=True,
          range_max=vocabulary_size,
          seed=None)
    
    ray= tf.fill( dims =[num_sampled] ,  value= (num_sampled/vocabulary_size)   ) 
    fillvalue2 = tf.cast(ray, testDataType)

    jay= tf.fill( dims =[ batch_size  ,1] ,  value= (num_sampled/vocabulary_size)   ) 
    fillvalue = tf.cast(jay, testDataType)

    true_expected_count2 = tf.get_variable( 'this1' ,dtype=testDataType, initializer = fillvalue  )
    sampled_expected_count2 = tf.get_variable( 'this2' , dtype=testDataType, initializer = fillvalue2   )

    sampled_value_16 = LogUniformCandidateSampler(
        sampled_values.sampled_candidates,
        true_expected_count2,
        sampled_expected_count2 )
```

You can switch between float16 and float32 by changing the `testDataType` variable at the top. You can also change the embedding size at the top by changing the `embedding_size` variable, which I recommend for switching between TPU mode and GPU mode in Google Colab. 

After you select which datatype and embedding size you want to use, select 'Runtime' -> 'Run all', and you can check the time it takes between steps at the training code at the bottom. 

In GPU mode, 100 steps takes about 11 seconds for float32, but takes 15 seconds for float16. I imagine float16 would have been faster. For GPU mode I used an embedding size of 52 (highest embedding size I could use before the system would crash during training for float32).

In TPU mode, 100 steps takes about 10 seconds for float32, but takes 6 minutes 5 seconds for float16. So 36x slower for float16 than float32. For TPU mode I used an embedding size of 154 (highest embedding size I could use before the system would crash during training for float32).

>OS Platform and Distribution

Python 3. Not sure about the rest, whatever Google Colab uses. 

>TensorFlow installed from

Not sure, whatever Google Colab uses. 

>CUDA/cuDNN version

Not sure, whatever Google Colab uses. 

>GPU model and memory

I believe Colab for GPU uses a Tesla k80, and for TPU they are using V2. 

>Exact command to reproduce

Same code as linked above 

https://drive.google.com/open?id=1bnlSfezPqtwapvURRTAqGR4mkcCl9QZl

>Mobile device

N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by Santosh-Gupta** [comment]
Here it is. I'll also edit this into the original post. 

>Have I written custom code

Here is the code I used, for convenience it's in a Google Colab notebook so you can just run the code in the notebook. 

https://drive.google.com/open?id=1bnlSfezPqtwapvURRTAqGR4mkcCl9QZl

I wrote a custom function to replace `uniform_candidate_sampler` because I wanted to isolate the issue if the slow down is due to using float16 instead of float32. The specific code to replace  `uniform_candidate_sampler` is

```
    LogUniformCandidateSampler = namedtuple("namedtuple", ["sampled_candidates", "true_expected_count", "sampled_expected_count"]) 
    sampled_values = tf.nn.uniform_candidate_sampler(
          true_classes=tf.cast(train_labels, tf.int64), num_sampled=num_sampled,
          num_true=1,
          unique=True,
          range_max=vocabulary_size,
          seed=None)
    
    ray= tf.fill( dims =[num_sampled] ,  value= (num_sampled/vocabulary_size)   ) 
    fillvalue2 = tf.cast(ray, testDataType)

    jay= tf.fill( dims =[ batch_size  ,1] ,  value= (num_sampled/vocabulary_size)   ) 
    fillvalue = tf.cast(jay, testDataType)

    true_expected_count2 = tf.get_variable( 'this1' ,dtype=testDataType, initializer = fillvalue  )
    sampled_expected_count2 = tf.get_variable( 'this2' , dtype=testDataType, initializer = fillvalue2   )

    sampled_value_16 = LogUniformCandidateSampler(
        sampled_values.sampled_candidates,
        true_expected_count2,
        sampled_expected_count2 )
```

You can switch between float16 and float32 by changing the `testDataType` variable at the time. You can also change the embedding size at the top by changing the `embedding_size` variable, which I recommend for switching between TPU mode and GPU mode in Google Colab. 

After you select which datatype and embedding size you want to use, select 'Runtime' -> 'Run all', and you can check the time it takes between steps at the training code at the bottom. 

In GPU mode, 100 steps takes about 11 seconds for float32, but takes 15 seconds for float16. I imagine float16 would have been faster. For GPU mode I used an embedding size of 52 (highest embedding size I could use before the system would crash during training).

In TPU mode, 100 steps takes about 10 seconds for float32, but takes 6 minutes 5 seconds for float16. So 36x slower for float16 than float32. For TPU mode I used an embedding size of 154 (highest embedding size I could use before the system would crash during training).

>OS Platform and Distribution

Not sure, whatever Google Colab uses. 

>TensorFlow installed from

Not sure, whatever Google Colab uses. 

>CUDA/cuDNN version

Not sure, whatever Google Colab uses. 

>GPU model and memory

I believe Colab for GPU uses a Tesla k80, and for TPU they are using V2. 

>Exact command to reproduce

Same code as linked above 

https://drive.google.com/open?id=1bnlSfezPqtwapvURRTAqGR4mkcCl9QZl

>Mobile device

N/A

**Comment by tensorflowbutler** [comment]
Nagging Assignee @harshini-gadige: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@Santosh-Gupta,
Can you please look into [Mixed Precision Guide](https://www.tensorflow.org/guide/mixed_precision) and check if it helps? Thanks!
